### ISSUE TITLE
rapidjson's default allocator does not handle allocation failures properly

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#8521]
Spin-off from #8515

Scylla uses the rapidjson library for parsing JSON documents. Unfortunately, as observed in Tencent/rapidjson#1168, Tencent/rapidjson#1359 and Tencent/rapidjson#1269, once the allocation fails due to not enough memory, rapidjson's default behavior is undefined - it does not check whether its allocator returned a proper address or null and proceeds, which usually results in a segmentation fault if we're lucky or memory corruption if we're less so.

One workaround to consider is overriding the default allocator to throw an exception once it's unable to allocate/reallocate memory. The request would still fail, but at least it would do so gracefully in a controlled fashion. Rapidjson's allocator is a concept implementing 3 methods:
https://rapidjson.org/classrapidjson_1_1_allocator.html
```cpp
concept Allocator {
    static const bool kNeedFree;    //!< Whether this allocator needs to call Free().
 
    // Allocate a memory block.
    // \param size of the memory block in bytes.
    // \returns pointer to the memory block.
    void* Malloc(size_t size);
 
    // Resize a memory block.
    // \param originalPtr The pointer to current memory block. Null pointer is permitted.
    // \param originalSize The current size in bytes. (Design issue: since some allocator may not book-keep this, explicitly pass to it can save memory.)
    // \param newSize the new size in bytes.
    void* Realloc(void* originalPtr, size_t originalSize, size_t newSize);
 
    // Free a memory block.
    // \param pointer to the memory block. Null pointer is permitted.
    static void Free(void *ptr);
};
```
It should be simple enough to provide a wrapper over the default allocator, but it should be preceded with code inspection and tests to ensure that throwing does not leave internal rapidjson objects in invalid state.

**Comment by avikivity** [comment]
Wow, this is terrible.

Combined with them refusing to make releases, I'm sorry we use them.

**Comment by psarna** [comment]
Yeah. On the other hand - they make it quite easy to hack through everything. Thankfully allocators are abstracted away and users can fix them on their own, which is still better than writing your own JSON parser.

**Comment by psarna** [comment]
(although, writing your own JSON parser *as a student project* is to be considered)

**Comment by psarna** [comment]
anecdote: I found a very interesting problem with `fmt` when implementing the allocator. I put the allocator in our internal `rjson` namespace and some random alternator tests started failing. After the gdb investigation it turned out that the error originates from `fmt::format`, which somehow calls `rjson::to_string_view` on an object which is not a string, which causes an assertion failure. Now, that's because `fmt::format` calls an unqualified `to_string_view` function, the compiler tries to look up the name in our `rjson` namespace and finds it - the identical name is absolutely incidental. It wasn't the case when we used the original allocator, which resided in an external `rapidjson` namespace, so the compiler wasn't able to look it up.
